hyppo/independence/hhg.py

```diff
@@ -129,16 +176,35 @@ def statistic(self, x, y):
         distx = x
         disty = y
 
-        if not self.is_distance:
-            distx, disty = compute_dist(
-                x, y, metric=self.compute_distance, **self.kwargs
-            )
-
-        S = _pearson_stat(distx, disty)
-        mask = np.ones(S.shape, dtype=bool)
-        np.fill_diagonal(mask, 0)
-        stat = np.sum(S[mask])
-        self.stat = stat
+        if not self.auto:
+            if not self.is_distance:
+                distx, disty = compute_dist(
+                    x, y, metric=self.compute_distance, **self.kwargs
+                )
+            S = _pearson_stat(distx, disty)
+            mask = np.ones(S.shape, dtype=bool)
+            np.fill_diagonal(mask, 0)
+            stat = np.sum(S[mask])
+            self.stat = stat
+
+        else:
+            # Fast HHG
+            if not self.is_distance:
+                pointer = (np.mean(x, axis=0), np.mean(y, axis=0))
+                zx, zy = pointer
+                zx = np.array(zx).reshape(1, -1)
+                zy = np.array(zy).reshape(1, -1)
+                xin = np.concatenate((zx, x))
+                yin = np.concatenate((zy, y))
+                distx, disty = compute_dist(
+                    xin, yin, metric=self.compute_distance, **self.kwargs
+                )
+                distx = distx[0]
```

Comment:
what are 202-205 doing?

Reply:
Due to how I cocatenate the center point as the first element of the sample point array, the first row of distance matrix returned is the distance of the other sample points from the center point - the 1D collection of distances I need to put into the univariate test - so long as I remove the 0.

Reply:
no the median over all dimensions?

Reply:
I may be wrong but doesn't compute_distance perform a pairwise distance calculation on a sample of points and the way the distance matrix is organized is a n x n matrix where ith-row, j-th column spot is the distance between the ith point and the j-th point? My logic was that if I made the center point the first point in the input array, then the first row (or column) will contain the distance of all sample points from the center. This is what I got from some testing with the function.

Unless the metric used is callable, in which case my knowledge ends there.

Reply:
A little more testing did show me that the distance between the center point and itself is not necessarily 0 depending on the metric, so I will have to just simply remove that 1st element in the distance matrix. But I won't do anything until we make sure I'm on the right track - just a note for myself.

Reply:
> I may be wrong but doesn't compute_distance perform a pairwise distance calculation on a sample of points and the way the distance matrix is organized is a n x n matrix where ith-row, j-th column spot is the distance between the ith point and the j-th point? My logic was that if I made the center point the first point in the input array, then the first row (or column) will contain the distance of all sample points from the center. This is what I got from some testing with the function.
> 
> Unless the metric used is callable, in which case my knowledge ends there.

I'm confused by this. Shouldn't this always be 0 for a distance matrix? Or 1 if we are looking at a kernel similarity matrix?

Reply:
If you mean the diagonal of the matrix, yeah, I believe you are correct! But isn't the ith row (or column) of the matrix also containing the distance of the other points from the ith point? So, given my target is a 1D collection of distances from the center point, my order of operations was (looking at just x here):

1) Add the center point as the first point of the sample point array, getting xin - with a shape of (n+1, p).
2) Put the new sample arrays into the compute_dist function, which gives distx - the pairwise distance matrix with shape (n+1, n+1).
3) Take the first row of distx - a collection of distances between the first point (the center) and the other sample points
4) Remove the element associated with the diagonal from the row - given its the first row, this is the first element.

I think this should be it?

Reply:
ah I get it, that is correct